components/at_mention/at_mention.tsx

```diff
@@ -1,41 +1,52 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
-import PropTypes from 'prop-types';
 import React from 'react';
 import {Overlay} from 'react-bootstrap';
 
 import {Client4} from 'mattermost-redux/client';
 import {displayUsername} from 'mattermost-redux/utils/user_utils';
+import {NameMappedObjects, UsernameMappedObjects} from 'mattermost-redux/types/utilities';
+import {UserProfile} from 'mattermost-redux/types/users';
+import {Group} from 'mattermost-redux/types/groups';
 
 import ProfilePopover from 'components/profile_popover';
 
-import {popOverOverlayPosition} from 'utils/position_utils.tsx';
+import {popOverOverlayPosition} from 'utils/position_utils';
+
 const spaceRequiredForPopOver = 300;
 
-export default class AtMention extends React.PureComponent {
-    static propTypes = {
-        children: PropTypes.node,
-        currentUserId: PropTypes.string.isRequired,
-        channelId: PropTypes.string,
-        hasMention: PropTypes.bool,
-        disableHighlight: PropTypes.bool,
-        disableGroupHighlight: PropTypes.bool,
-        isRHS: PropTypes.bool,
-        mentionName: PropTypes.string.isRequired,
-        teammateNameDisplay: PropTypes.string.isRequired,
-        usersByUsername: PropTypes.object.isRequired,
-        groupsByName: PropTypes.object.isRequired,
-    };
-
-    static defaultProps = {
+type Props = {
+    children?: React.ReactNode;
+    currentUserId: string;
+    channelId?: string;
+    hasMention?: boolean;
+    disableHighlight?: boolean;
+    disableGroupHighlight?: boolean;
+    isRHS?: boolean;
+    mentionName: string;
+    teammateNameDisplay: string;
+    usersByUsername: UsernameMappedObjects<UserProfile>;
+    groupsByName: NameMappedObjects<Group>;
+}
+
+type State = {
+    show: boolean;
+    target?: EventTarget;
```

Comment:
It looks a little weird to see this as `EventTarget` when I feel like it should be a DOM node. If you specify the type of the event passed to `handleClick` as `React.MouseEvent<HTMLAnchorElement>` to match the type given to the ref, does that let you define target as an `HTMLAnchorElement` and avoid the `as` when passing `this.state.target` into `Overlay` as a prop?

Reply:
even with defining `React.MouseEvent<HTMLAnchorElement>` as props type for `handleClick` i had to cast `e.target` as `HTMLAnchorElement` since `React.MouseEvent` extends `BaseSyntheticEvent` and the event type in `BaseSyntheticEvent` is `EventTarget & T`.

Reply:
Oh, I see. I think I like having the `as HTMLAnchorElement` in the click handler like you've done now though, so this looks good to me